### DISCUSSION TITLE
`@alpinejs/focus` Configurable `displayCheck` setting

### DISCUSSION THREAD

**Discussion by joshuapease** [alpinejs/alpine#4390] [3. Feature ideas]
I've run into some issues building a [disclosure navigation](https://www.w3.org/WAI/ARIA/apg/patterns/disclosure/examples/disclosure-navigation) and using the focus plugin.

Child menu items are hidden using `x-show` (making them `display: none`).

When I use `@keydown.down="$focus.wrap().next()"` and `@keydown.up="$focus.wrap().previous()"` focus halts when the next focusable element is hidden. 

My best guess is that [tabbable](https://github.com/focus-trap/tabbable) is finding the hidden element, but is unable to focus on it since it's hidden.

If I manually modify the plugin to use `displayCheck: 'full'` I get the behavior I'd expect.

![CleanShot 2024-10-10 at 10 15 53](https://github.com/user-attachments/assets/27ad38c6-9ddc-4175-b2ae-5cc747f35fcd)

Here's the spot I would modify.

https://github.com/alpinejs/alpine/blob/ab743bc5668f018db5d410cad4e48b755696ad9e/packages/focus/src/index.js#L36-L40

It seems like other folks have encountered this too:
- https://github.com/alpinejs/alpine/discussions/4228

**Comment by robsontenorio** [comment]
@joshuapease Have you found a solution?

**Comment by SpyridonLaz** [comment]
bump. this should be fixed as it makes searchable boxes a nightmare

**Comment by joshuapease** [comment]
Hey everyone! Been busy with client projects, but had a bit of free time to dig back in to this. 

I've been experimenting with two ways of customizing displayCheck. 

I have some Draft PRs with code, but would love to discuss this with the Alpine JS maintainers before actually opening PRs in this repo (there might be better ways of doing this). 

**Option 1** - Make the focus plugin a factory function. 

- https://github.com/joshuapease/alpine/pull/2

This allows us to pass a config object to the plugin. 

I tried to write this in a way that also maintains backwards compatibility.  

```
// Existing code - no changes needed
Alpine.plugin(focus)

// New configuration options
Alpine.plugin(focus({ displayCheck: 'full' }))    // Include hidden elements
Alpine.plugin(focus({ displayCheck: 'none' }))    // Skip hidden elements (default)
```

**Option 2** - `displayCheck()` chainable method.  

- https://github.com/joshuapease/alpine/pull/1

Takes inspiration from `$focus.within($myref).first()`. 

Chaining `.displayCheck('full')` enables that settings for your specific call. 

**Usage:**

$focus.displayCheck('full') is chained with your function calls.

```js
// Examples
this.$focus.displayCheck('full').next()
this.$focus.displayCheck('full').previous()
```

---

Would love to hear from the maintainers of Alpine if this is a customization we could add to the focus plugin. It unlocks a lot of possibilities, especially when dealing with hidden nav, dropdowns, accordions groups, etc.

**Reply by jonnitto** [reply]
I'm not a maintainer, but if I could choose one, I would prefer option 2, as it allows to set it per usage. But why not add both, so you could set a default behavior and override if needed?
